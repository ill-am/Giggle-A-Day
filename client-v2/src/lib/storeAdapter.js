import { writable } from "svelte/store";

// Minimal in-memory adapter: exposes a Svelte writable store and a factory
export function createStoreAdapter(initial = "") {
  const store = writable(initial);

  // Attach a simple id for diagnostics (optional)
  store.__chronos_id = "client-v2-preview-store";
  store.__is_canonical = true;

  return store;
}

// Create and export a singleton previewStore for convenience tests
export const previewStore = createStoreAdapter("");

export default {
  createStoreAdapter,
  previewStore,
};
