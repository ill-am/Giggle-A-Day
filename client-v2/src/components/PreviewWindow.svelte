<script>
  import { previewStore } from "../lib/storeAdapter.js";
  import { onMount } from "svelte";

  // Minimal uiState for showing loading state; in real implementation
  // this will be wired to a proper store.
  export let uiState = { status: "idle", message: "" };

  // Spinner (inline small component)
  let showSkeleton = false;
  let skeletonTimer;
  const SKELETON_MIN_MS = 200; // ensure short visibility to avoid flicker

  $: if (uiState.status === "loading") {
    clearTimeout(skeletonTimer);
    showSkeleton = true;
  } else {
    // keep skeleton visible for a short minimum time to avoid flicker
    clearTimeout(skeletonTimer);
    skeletonTimer = setTimeout(() => {
      showSkeleton = false;
    }, SKELETON_MIN_MS);
  }

  // Mirror test hook behavior: update body attribute when preview present
  $: if (typeof document !== "undefined") {
    if ($previewStore && $previewStore.length > 0) {
      document.body.setAttribute("data-preview-ready", "1");
    } else {
      document.body.removeAttribute("data-preview-ready");
    }
  }

  // Dev debug overlay toggle
  let showDebug = false;
  onMount(() => {
    try {
      const params =
        typeof window !== "undefined"
          ? new URLSearchParams(window.location.search)
          : null;
      if (params && params.get("debugPreview") === "1") showDebug = true;
      if (
        typeof window !== "undefined" &&
        window.localStorage &&
        window.localStorage.getItem("__SHOW_PREVIEW_DEBUG__") === "1"
      )
        showDebug = true;
    } catch (e) {}
  });
</script>

<div class="preview-container">
  {#if uiState.status === "loading"}
    <div class="loading-overlay">
      {#if showSkeleton}
        <div class="skeleton">
          <div class="s-line" style="width:70%"></div>
          <div class="s-line" style="width:90%"></div>
          <div class="s-line" style="width:60%"></div>
        </div>
      {:else}
        <p>{uiState.message || "Loading Preview..."}</p>
      {/if}
    </div>
  {:else if $previewStore}
    <div class="preview-content" data-testid="preview-content">
      {@html $previewStore}
    </div>
  {:else}
    <div class="placeholder">Your generated preview will appear here.</div>
  {/if}
</div>

{#if showDebug}
  <div class="preview-debug-overlay" data-testid="preview-debug-overlay">
    <h4>Preview Debug</h4>
    <div>
      <strong>Preview length:</strong>
      {$previewStore ? $previewStore.length : 0}
    </div>
    <div style="margin-top:8px">
      <strong>Preview HTML (truncated):</strong>
      <pre>{String($previewStore || "").substring(0, 2000)}</pre>
    </div>
  </div>
{/if}

<style>
  .preview-container {
    position: relative;
    height: 100%;
    width: 100%;
    border: 1px solid #e0e0e0;
    border-radius: 8px;
    background-color: #f9f9f9;
    overflow-y: auto;
    min-height: 240px;
    display: flex;
    justify-content: center;
    align-items: center;
    text-align: center;
  }
  .loading-overlay {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    height: 100%;
    color: #888;
  }
  .preview-content {
    padding: 1.5rem;
    text-align: left;
    width: 100%;
    height: 100%;
  }
  .skeleton {
    display: flex;
    flex-direction: column;
    gap: 8px;
  }
  .s-line {
    height: 12px;
    background: #eee;
    border-radius: 6px;
  }
  .preview-debug-overlay {
    position: fixed;
    right: 12px;
    bottom: 12px;
    width: 320px;
    max-height: 40vh;
    overflow: auto;
    background: rgba(0, 0, 0, 0.8);
    color: #fff;
    font-family: monospace;
    font-size: 12px;
    padding: 8px;
    border-radius: 6px;
    z-index: 9999;
    box-shadow: 0 6px 18px rgba(0, 0, 0, 0.3);
  }
  /* Print-safe rules: avoid breaking preview content across pages */
  @media print {
    .preview-content {
      page-break-inside: avoid;
    }
    .preview-container {
      border: none;
      background: #fff;
    }
  }

  /* Use a readable serif/sans stack similar to canonical client */
  .preview-content,
  .preview-container {
    font-family: Georgia, "Times New Roman", Times, serif;
    color: #111;
  }
</style>
