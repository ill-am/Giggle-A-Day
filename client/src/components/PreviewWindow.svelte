<script>
  import { contentStore, previewStore, uiStateStore, setUiLoading, setUiSuccess, setUiError } from '../stores';
  import { loadPreview } from '../lib/api';
  import Spinner from './Spinner.svelte';
  import PreviewSkeleton from './PreviewSkeleton.svelte';
  import { onMount } from 'svelte';

  import { debounce } from '../lib/utils';

  let content;
  contentStore.subscribe(value => {
    content = value;
    if (content && autoPreview) {
      // Debounced auto update to avoid rapid requests
      debouncedUpdate(content);
    }
  });

  // Instrument previewStore updates using Svelte's auto-subscription ($previewStore)
  // reactive local preview HTML used by the template to avoid any indirect render races
  let lastPreview = '';
  let previewHtmlLocal = '';

  $: if (typeof $previewStore !== 'undefined' && $previewStore !== lastPreview) {
    const value = $previewStore;
    previewHtmlLocal = value || '';
    console.log('PreviewWindow: previewStore updated, length=', value ? value.length : 0);
    // set DOM-visible attribute for tests when preview is present and include a timestamp
    try {
      if (typeof document !== 'undefined' && document.body) {
        const ts = String(Date.now());
        // mark body for backward compatibility
        try { document.body.setAttribute('data-preview-ready', value ? '1' : '0'); } catch (e) {}
        try { document.body.setAttribute('data-preview-timestamp', value ? ts : ''); } catch (e) {}
        try { window.dispatchEvent(new CustomEvent('preview-ready', { detail: { timestamp: ts } })); } catch (e) {}
        setTimeout(() => { try { document.body.removeAttribute('data-preview-ready'); document.body.removeAttribute('data-preview-timestamp'); } catch (e) {} }, 8000);
        // also mark the preview-content element directly so automated checks targeting it succeed
        try {
          const el = document.querySelector('[data-testid="preview-content"]');
          if (el) {
            el.setAttribute('data-preview-ready', value ? '1' : '0');
            if (value) el.setAttribute('data-preview-timestamp', ts);
            else el.removeAttribute('data-preview-timestamp');
          }
        } catch (e) {}
        // expose last preview HTML to tests
        try { if (typeof window !== 'undefined') window['__LAST_PREVIEW_HTML'] = value; } catch (e) {}
        // ensure UI state and local flag reflect the preview presence so the template renders
        try { uiStateStore.set({ status: 'success', message: 'Preview loaded' }); } catch (e) {}
      }
    } catch (e) {}
    lastPreview = value;
  }

  // computed reactive flag: true when previewStore contains non-empty HTML
  $: computedHasPreview = previewHtmlLocal && String(previewHtmlLocal).trim().length > 0;

  let uiState = { status: 'idle', message: '' };
  uiStateStore.subscribe(value => {
    uiState = value || { status: 'idle', message: '' };
  });

  // expose debug hook for automated verification: latest preview HTML
  let latestPreviewHtml = '';
  $: if ($previewStore) {
    latestPreviewHtml = $previewStore;
    try { console.debug('[DEV] PreviewWindow: $previewStore length=', String($previewStore).length); } catch (e) {}
    try {
      // @ts-ignore
      window.__preview_updated_ts = Date.now();
      // @ts-ignore
      window.__preview_html_snippet = String($previewStore).slice(0, 1200);
    } catch (e) {}
  }

  // Dev-only DOM marker to help automated verification detect updates
  $: if (import.meta.env.DEV && $previewStore) {
    try {
      const el = document.querySelector('.preview-container');
      if (el) el.setAttribute('data-preview-updated', String(Date.now()));
    } catch (e) {}
  }

  // Background preview URL derived from content.background (filename or absolute URL)
  $: bgUrl = null;
  $: if (content && content.background) {
    if (typeof content.background === 'string' && content.background.startsWith('http')) {
      bgUrl = content.background;
    } else {
      bgUrl = `/samples/images/${encodeURIComponent(content.background)}`;
    }
  }

  // Preview controls
  let autoPreview = true;
  let flash = false;

  // Build a tiny client-side preview HTML (safe-escaped) to use as a fallback
  const escapeHtml = (str) => {
    if (!str) return '';
    return String(str)
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&#39;');
  };

  const buildLocalPreviewHtml = (content) => {
    const title = escapeHtml(content.title || 'Preview');
    const body = escapeHtml(content.body || '');
    return `\n      <article class="local-preview-fallback" style="padding:1.25rem">\n        <h2 style=\"margin-top:0;\">${title}</h2>\n        <div>${body.replace(/\n/g, '<br/>')}</div>\n      </article>\n    `;
  };

  const updatePreview = async (newContent) => {
    if (!newContent) {
      previewStore.set('');
      return;
    }
    try {
      uiStateStore.set({ status: 'loading', message: 'Loading preview...' });
      const html = await loadPreview(newContent);
      previewStore.set(html);
      // trigger brief flash to draw attention
      flash = true;
      setTimeout(() => (flash = false), 600);
      uiStateStore.set({ status: 'success', message: 'Preview loaded' });
    } catch (error) {
      uiStateStore.set({ status: 'error', message: `Failed to load preview: ${error.message}` });
      previewStore.set('');
    }
  };

  // Debounced auto update to avoid rapid requests (200ms for snappier feel)
  const debouncedUpdate = debounce(updatePreview, 200);

  onMount(() => {
    if (content) {
      updatePreview(content);
    }
  });
</script>

  <div class="preview-container">
  <div class="preview-controls">
    <label><input type="checkbox" data-testid="auto-preview-checkbox" bind:checked={autoPreview} /> Auto-preview</label>
    <button data-testid="preview-now-button" on:click={() => updatePreview(content)} disabled={!content || uiState.status === 'loading'}>
      {#if uiState.status === 'loading'}
        Previewing...
      {:else}
        Preview Now
      {/if}
    </button>
  </div>

  {#if uiState.status === 'loading'}
    <div class="loading-overlay">
      <p>Loading Preview...</p>
    </div>
  {:else if $previewStore}
    <div class="preview-stage {flash ? 'flash' : ''}">
      {#if bgUrl}
        <div class="bg-preview"><img src={bgUrl} alt="background preview" /></div>
      {/if}
      <div class="preview-content" data-testid="preview-content">
        {@html $previewStore}
      </div>
    </div>
  {:else}
    {#if content}
      <!-- Fallback: render a minimal, safe client-side preview from content
           This ensures the preview pane reflects generated content even when
           server-side preview fails or the proxy blocks the request. -->
      <div class="preview-stage fallback-stage">
        <div class="preview-content" data-testid="preview-content">{@html buildLocalPreviewHtml(content)}</div>
      </div>
    {:else}
      <div class="placeholder">
        <p>Your generated preview will appear here.</p>
      </div>
    {/if}
  {/if}
</div>

<style>
  .preview-container {
    position: relative;
    height: 100%;
    width: 100%;
    border: 1px solid #e0e0e0;
    border-radius: 8px;
    background-color: #f9f9f9;
    overflow-y: auto;
    min-height: 240px; /* ensure preview area isn't collapsed to zero height */
  }
  .loading-overlay {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100%;
    color: #888;
  }
  .preview-content {
    padding: 1.5rem;
    text-align: left;
  }

  .preview-stage { position: relative; min-height: 300px }
  .bg-preview { position: absolute; inset: 0; opacity: 0.45; pointer-events: none }
  .bg-preview img { width: 100%; height: 100%; object-fit: cover }
  .preview-stage .preview-content { position: relative; z-index: 2 }

  .small-spinner {
    width: 28px;
    height: 28px;
    border-radius: 50%;
    border: 4px solid rgba(0,0,0,0.08);
    border-top-color: #2c3e50;
    animation: spin 0.9s linear infinite;
    position: absolute;
    top: 12px;
    right: 12px;
    z-index: 10;
  }

  @keyframes spin { from { transform: rotate(0deg) } to { transform: rotate(360deg) } }

  /* flash highlight when preview updates */
  .preview-stage.flash {
    box-shadow: 0 0 0 3px rgba(66, 153, 225, 0.15) inset, 0 0 0 2px rgba(66,153,225,0.08);
    transition: box-shadow 0.45s ease-out;
  }
</style>
