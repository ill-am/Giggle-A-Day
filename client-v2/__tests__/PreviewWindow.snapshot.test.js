import { describe, it, expect } from "vitest";
import { render } from "@testing-library/svelte/svelte5";
import PreviewWindow from "../src/components/PreviewWindow.svelte";
import { previewStore } from "../src/lib/storeAdapter.js";

describe("PreviewWindow snapshot", () => {
  it("matches DOM snapshot with sample content", () => {
    previewStore.set("<h2>Snapshot Poem</h2><p>Snapshot line</p>");
    const { container } = render(PreviewWindow, {
      uiState: { status: "idle" },
    });
    // Use a simple DOM snapshot to capture structure
    expect(container.innerHTML).toMatchSnapshot();
  });
});
