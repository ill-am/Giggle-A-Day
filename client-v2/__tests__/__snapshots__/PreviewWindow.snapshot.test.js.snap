// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PreviewWindow snapshot > matches DOM snapshot with sample content 1`] = `"<div class="preview-container svelte-1xfprg0"><!----><div class="preview-content svelte-1xfprg0" data-testid="preview-content"><h2>Snapshot Poem</h2><p>Snapshot line</p><!----></div><!----></div> <!---->"`;
